doc/sphinx/proofs/writing-proofs/rewriting.rst

```diff
@@ -1,102 +1,126 @@
 
-      Uses the equality :n:`@term`:sub:`1` :n:`= @term` :sub:`2` from right to left
+      :n:`{? forall {+ (x__i: A__i) } , } EQ @term__1 @term__2`
 
-   .. tacv:: rewrite @term in @goal_occurrences
+   where :g:`EQ` is the Leibniz equality `eq` or a registered setoid equality.
+   Note that :n:`eq @term__1 @term__2` is typically written with the infix notation
+   :n:`@term__1 = @term__2`.  You must `Require Setoid` to use the tactic
+   with a setoid equality or with setoid rewriting.  In the general form,
+   any :n:`@binder` may be used, not just :n:`(x__i: A__i)`.
 
-      Analogous to :n:`rewrite @term` but rewriting is done following
-      the clause :token:`goal_occurrences`. For instance:
+   .. todo doublecheck the @binder comment is correct.
 
-      + :n:`rewrite H in H'` will rewrite `H` in the hypothesis
-        ``H'`` instead of the current goal.
-      + :n:`rewrite H in H' at 1, H'' at - 2 |- *` means
-        :n:`rewrite H; rewrite H in H' at 1; rewrite H in H'' at - 2.`
-        In particular a failure will happen if any of these three simpler tactics
-        fails.
-      + :n:`rewrite H in * |-` will do :n:`rewrite H in H'` for all hypotheses
-        :g:`H'` different from :g:`H`.
-        A success will happen as soon as at least one of these simpler tactics succeeds.
-      + :n:`rewrite H in *` is a combination of :n:`rewrite H` and :n:`rewrite H in * |-`
-        that succeeds if at least one of these two tactics succeeds.
+   :n:`rewrite @one_term` finds subterms matching :n:`@term__1` in the goal,
+   and replaces them with :n:`@term__2` (or the reverse if `<-` is given).
+   Some of the variables :g:`x`\ :sub:`i` are solved by unification,
+   and some of the types :n:`A__1, ..., A__n` may become new
+   subgoals.  It won't find occurrences inside `forall` that refer
+   to variables bound by the `forall`; use :tacn:`setoid_rewrite`
+   if you want to find such occurrences.
 
-      Orientation :g:`->` or :g:`<-` can be inserted before the :token:`term` to rewrite.
+   :n:`{+, @oriented_rewriter }`
+     The :n:`@oriented_rewriter`\s are applied sequentially
+     to the first goal generated by the previous :n:`@oriented_rewriter`.  If any of them fail,
+     the tactic fails.
 
-   .. tacv:: rewrite @term at @occurrences
+   :n:`{? {| -> | <- } }`
+     For `->` (the default), :n:`term__1` is rewritten
+     into :n:`term__2`.  For `<-`, :n:`term__2` is rewritten into :n:`term__1`.
 
-      Rewrite only the given :token:`occurrences` of :token:`term`. Occurrences are
-      specified from left to right as for pattern (:tacn:`pattern`). The rewrite is
-      always performed using setoid rewriting, even for Leibniz’s equality, so one
-      has to ``Import Setoid`` to use this variant.
+   :n:`{? @natural } {? {| ? | ! } }`
+     :n:`natural` is the number of rewrites to perform.  If `?` is given, :n:`@natural`
+     is the maximum number of rewrites to perform; otherwise :n:`natural` is the exact number
```

Comment:
```suggestion
     :n:`@natural` is the number of rewrites to perform.  If `?` is given, :n:`@natural`
     is the maximum number of rewrites to perform; otherwise :n:`@natural` is the exact number
```